scala_proto/private/scalapb_aspect.bzl

```diff
@@ -2,255 +2,162 @@ load("@rules_proto//proto:defs.bzl", "ProtoInfo")
-
-        code_generator = toolchain.code_generator
-
-        if compile_protos:
-            scalapb_file = ctx.actions.declare_file(
-                target.label.name + "_scalapb.srcjar",
-            )
-            proto_to_scala_src(
-                ctx,
-                target.label,
-                code_generator,
-                compile_protos,
-                transitive_protos,
-                target_ti.transitive_proto_path.to_list(),
-                flags,
-                scalapb_file,
-                toolchain.named_generators,
-                sorted(extra_generator_jars),
-            )
-
-            src_jars = depset([scalapb_file])
-            output = _compiled_jar_file(ctx.actions, scalapb_file)
-            outs = depset([output])
-            java_info = _compile_scala(
-                ctx,
-                toolchain.scalac,
-                target.label,
-                output,
-                scalapb_file,
-                deps,
-                imps,
-                compile_protos,
-                "" if target_ti.proto_source_root == "." else target_ti.proto_source_root,
-            )
-        else:
-            # this target is only an aggregation target
-            src_jars = depset()
-            outs = depset()
-            java_info = java_common.merge(_concat_lists(deps, imps))
-
-    return [
-        ScalaPBAspectInfo(
-            src_jars = src_jars,
-            output_files = outs,
-            proto_info = transitive_ti,
-            java_info = java_info,
-        ),
-    ]
+    toolchain = ctx.toolchains["@io_bazel_rules_scala//scala_proto:toolchain_type"]
+    proto = target[ProtoInfo]
+    deps = [d[ScalaPBAspectInfo].java_info for d in ctx.rule.attr.deps]
 
-def _concat_lists(list1, list2):
-    all_providers = []
-    all_providers.extend(list1)
-    all_providers.extend(list2)
-    return all_providers
+    if proto.direct_sources and _code_should_be_generated(ctx, toolchain):
+        src_jars = _generate_sources(ctx, toolchain, proto)
+        java_info = _compile_sources(ctx, toolchain, proto, src_jars, deps)
+        return [ScalaPBAspectInfo(java_info = java_info)]
+    else:
+        # this target is only an aggregation target
+        return [ScalaPBAspectInfo(java_info = java_common.merge(deps))]
 
 scalapb_aspect = aspect(
     implementation = _scalapb_aspect_impl,
     attr_aspects = ["deps"],
-    required_aspect_providers = [
-        [ProtoInfo],
-        [ScalaPBImport],
-    ],
+    incompatible_use_toolchain_transition = True,
     attrs = {
-        "_protoc": attr.label(
-            executable = True,
-            cfg = "exec",
-            default = "@com_google_protobuf//:protoc",
+        "_java_toolchain": attr.label(
+            default = Label("@bazel_tools//tools/jdk:current_java_toolchain"),
+        ),
+        "_host_javabase": attr.label(
+            default = Label("@bazel_tools//tools/jdk:current_java_runtime"),
+            cfg = "host",
```

Comment:
should be exec?

Reply:
I think not. It's same as https://github.com/bazelbuild/rules_scala/blob/5aecfd9b1b06d22ae549cd96101ec1242b7566bf/scala/private/common_attributes.bzl#L95-L98 
I assume `_host_javabase` implies `cfg=host`

Reply:
Changed `cfg` to `exec`